Add support for debug modes

pinzart opened the pull request:
### Purpose

Add support for debug modes (similar to feature flags).
The main idea behind this PR is to allow any developer to wrap functionality under flags. The code can get pushed to master...and not interfere with other people's work. It can even get shipped to users. As long as the "experimental code" is wrapped under a DebugMode and not activated...it should not change any of the existing functionality.

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

(FILL ME IN) Reviewer 1  (If possible, assign the Reviewer for the PR)

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

pinzart commented:
Still need to add UI under Debug tab.

pinzart commented on src/DynamoCore/App.config:
```diff
@@ -1,5 +1,9 @@
+   <configSections>
+      <section name="DebugModes" type="System.Configuration.AppSettingsSection"/>
+   </configSections>
+
```
To add enable a debug mode (from config file)...one would have to add a new entry under a DebugModes section like so:
`
   <DebugModes>
      <add key="DebugMode1" value="true"/>
      <add key="DebugMode2" value="true"/>
   </DebugModes>
`

pinzart replied:
I could add a separate config file...(debug.config) just so that we do not have to pollute the DynamoCore config ...

mjkkirschner commented:
Hi @pinzart a few questions about the context of this PR:
What is work does this this pull request support? 
What modes do you wish to enable or disable via config file? 
What about the existing debug menu does not satisfy the needs of your work?

QilongTang commented:
@pinzart For local testing seems the current debug mode is enough unless you plan to distribute Dynamo in such a way?

pinzart commented:
_What work does this this pull request support?_
@mjkkirschner The main idea behind this PR is to allow developers to wrap functionality under flags (basically feature flags). THe code can get pushed to master...and not interfere with other people's work. It can even get shipped to users. As long as the "experimental code" is wrapped under a DebugMode and not activated...it should not change any of the existing functionality.

_What modes do you wish to enable or disable via config file?_
I am hoping to add Debug modes for ADP (ex: EnableADPAnalytics, EnableADPOptinPage etc)

What about the existing debug menu does not satisfy the needs of your work?
THe debug menu is good for a new UI (a list of available debugmodes ...displayed as checkboxes + name + description). Exactly like Revit has...

pinzart commented:
@QilongTang 
_For local testing seems the current debug mode is enough unless you plan to distribute Dynamo in such a way?_
This PR does not add any debug mode. It is just the infrastructure to support adding debug modes. By local testing.,..you mean testing the infrastructure code of the current PR...or the functionality behind new debug modes that a developer could add down the line ?

pinzart commented:
This is what the UI would look like
![image](https://user-images.githubusercontent.com/26064027/80423198-980e1180-88ad-11ea-8ff4-872a126c0674.png)

pinzart commented:
@mjkkirschner @QilongTang any more questions/thoughts on this ?

mjkkirschner commented:
@pinzart I still don't understand the relationship between this feature and the debug menu we already have.

Are you going to unify them? I much prefer not to have too competing debug feature flag menu options...

pinzart commented:
Here is how I merged the DebugModes with the DebugMenu
![image](https://user-images.githubusercontent.com/26064027/80428667-651d4b00-88b8-11ea-89b5-5654363e5bb6.png)

DebugTab -> DebugMenu -> DebugModesMenuItem => DebugModesWindow

pinzart commented:
@mjkkirschner let's talk tomorrow so I can better explain my changes

pinzart commented:
Added unit tests and switched to basic xml configuration file (makes it easier to test and handle).
debug.config should not be delivered as part of any release . It only makes sense to manually add it when you want to enable debug modes on startup.

QilongTang commented on src/DynamoCoreWpf/Views/Core/DynamoView.xaml:
```diff
@@ -814,7 +814,13 @@
+                                  Name="DebugModes"
+                                  Header="{x:Static p:Resources.DynamoViewDebugMenuDebugModes}"
+                                  Click="OnDebugModesClick"
+                                  DataContext="{Binding Path=Model.UpdateManager}"
```
This does not look like the correct DataContext?

pinzart replied:
nice catch

pinzart commented:
@mjkkirschner @QilongTang any more issues with this PR ?

QilongTang commented:
LGTM
